src/backend/distributed/connection/locally_reserved_shared_connection_stats.c

```diff
@@ -0,0 +1,487 @@
+								"disable reserved connection counters")));
+	}
+
+	entry->reservedConnectionCount++;
+
+	ereport(DEBUG5, (errmsg("Incremented shared reserved connection to node "
+							"%s:%d to %d", entry->key.hostname, entry->key.port,
+							entry->reservedConnectionCount)));
+}
+
+
+/*
+ * This function is intended to be called when the session has reserved connections
+ * but not used them, and we are in a good position to get rid of them.
+ */
+void
+DeallocateAllReservedConnections(void)
+{
+	HASH_SEQ_STATUS status;
+	ReservedConnectionCounterHashEntry *entry;
+
+	hash_seq_init(&status, SessionLocalReservedConnectionCounters);
+	while ((entry = (ReservedConnectionCounterHashEntry *) hash_seq_search(&status)) != 0)
+	{
+		while (entry->reservedConnectionCount > 0)
+		{
+			DecrementSharedConnectionCounter(entry->key.hostname, entry->key.port);
+			entry->reservedConnectionCount--;
+
+			ereport(DEBUG5, (errmsg("Decrement shared reserved connection to node "
+									"%s:%d to %d", entry->key.hostname, entry->key.port,
+									entry->reservedConnectionCount)));
+		}
+
+		bool found = false;
+
+		/*
+		 * We cleaned up all the entries because we may not need reserved connections
+		 * in the next iteration.
+		 */
+		hash_search(SessionLocalReservedConnectionCounters, entry, HASH_REMOVE, &found);
+		Assert(found);
+	}
+}
+
+
+/*
+ * DecrementReservedConnection decrements the reserved counter for the given host.
+ */
+void
+DecrementReservedConnection(const char *hostName, int nodePort, Oid databaseOid)
+{
+	ReservedConnectionCounterHashKey key;
+
+	strlcpy(key.hostname, hostName, MAX_NODE_LENGTH);
+	key.port = nodePort;
+	key.databaseOid = databaseOid;
+
+	bool found = false;
+	ReservedConnectionCounterHashEntry *entry =
+		(ReservedConnectionCounterHashEntry *) hash_search(
+			SessionLocalReservedConnectionCounters, &key, HASH_FIND, &found);
+
+	if (!found)
+	{
+		ereport(ERROR, (errmsg("BUG: untracked reserved connection"),
+						errhint("Set citus.max_shared_pool_size TO -1 to "
+								"disable reserved connection counters")));
+	}
+
+	entry->reservedConnectionCount--;
+
+	ereport(DEBUG5, (errmsg("Decrement shared reserved connection to node %s:%d to %d",
+							hostName, nodePort, entry->reservedConnectionCount)));
+}
+
+
+/*
+ * ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded is a wrapper around
+ * ReserveSharedConnectionCounterForNodeListIfNeeded.
+ */
+void
+ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded(void)
+{
+	List *primaryNodeList = ActivePrimaryNonCoordinatorNodeList(NoLock);
+
+	ReserveSharedConnectionCounterForNodeListIfNeeded(primaryNodeList);
+}
+
+
+/*
+ * ReserveSharedConnectionCounterForNodeListIfNeeded reserves a shared connection
+ * counter per node in the nodeList unless:
+ *  - there is at least one connection to the node
+ *  - has already reserved a connection to the node
+ */
+static void
+ReserveSharedConnectionCounterForNodeListIfNeeded(List *nodeList)
```

Comment:
I think this method is doing a bit too much. In general I find `IfNeeded` prefixes a bit incorrect because that means the function is doing at least two different things such as checking some conditions and doing some stuff.

For example:
- We can have a method to check if we can reserve any connections(having the if checks there)
- We can have a method to filter the nodes that have a connection or reserved a connection
- We can then have a method that takes the filtered nodes and just reserved the connections to them so it does only one thing

I will give a prototype:

```c
/*
 * EnsureEachPrimaryNodeHasReservedConnection reserves a connection
 * for each primary node if it doesn't already have one.
 */
void
EnsureEachPrimaryNodeHasReservedConnection(void)
{
	if (!IsReservationRequired()) {
		return;
	}
	List *primaryNodeList = ActivePrimaryNonCoordinatorNodeList(NoLock);
	/*
	 * We sort the workerList because adaptive connection management
	 * (e.g., OPTIONAL_CONNECTION) requires any concurrent executions
	 * to wait for the connections in the same order to prevent any
	 * starvation. If we don't sort, we might end up with:
	 *      Execution 1: Get connection for worker 1, wait for worker 2
	 *      Execution 2: Get connection for worker 2, wait for worker 1
	 *
	 *  and, none could proceed. Instead, we enforce every execution establish
	 *  the required connections to workers in the same order.
	 */
	primaryNodeList = SortList(primaryNodeList, CompareWorkerNodes);
	List * remainingNodeList = FilterOutNodesWithAvailableConnection(primaryNodeList);
	ReserveSharedConnectionForNodeList(remainingNodeList);
}

static bool IsReservationRequired() {
	if (GetMaxSharedPoolSize() == DISABLE_CONNECTION_THROTTLING)
	{
		/* connection throttling disabled */
		return false;
	}

	if (SessionLocalReservedConnectionCounters == NULL)
	{
		/*
		 * This is unexpected as SessionLocalReservedConnectionCounters hash table is
		 * created at startup. Still, let's be defensive.
		 */
		ereport(ERROR, (errmsg("unexpected state for session level reserved "
							   "connection counter hash")));
	}
	return true;
}

static List* FilterOutNodesWithAvailableConnection(const List* nodeList) {
	char *databaseName = get_database_name(MyDatabaseId);

	WorkerNode *workerNode = NULL;
	List* remainingNodeList = NIL;
	foreach_ptr(workerNode, nodeList)
	{
		if (ConnectionAvaliableToNode(workerNode->workerName, workerNode->workerPort,
									  CurrentUserName(), databaseName))
		{
			/*
			 * The same user has already an active connection for the node. It
			 * means that the execution can use the same connection, so reservation
			 * is not necessary.
			 */
			continue;
		}

		if (HasAlreadyReservedConnection(workerNode->workerName, workerNode->workerPort,
										 MyDatabaseId))
		{
			/*
			 * We already locally reserved a connection for this node. So, it is not
			 * necessary to reserve again.
			 *
			 * One caveat here is that we are ignoring for which user we reserved the
			 * connection. It is not important as Citus doesn't support switching
			 * between users (https://github.com/citusdata/citus/pull/3869) and
			 * we're clearing the reserved connections at the end of the transaction.
			 */
			continue;
		}
		remainingNodeList = lappend(remainingNodeList, workerNode);
	}
	return remainingNodeList;
}


/*
 * ReserveSharedConnectionForNodeListIfNeeded reserves a shared connection
 * per node in the nodeList
 */
static void
ReserveSharedConnectionForNodeList(List *nodeList)
{
	WorkerNode *workerNode = NULL;
	foreach_ptr(workerNode, nodeList)
	{
		/*
		 * We are trying to be defensive here by ensuring that the required hash
		 * table entry can be allocated. The main goal is that we don't want to be
		 * in a situation where shared connection counter is incremented but not
		 * the local reserved counter due to out-of-memory.
		 *
		 * Note that shared connection stats operate on the shared memory, and we
		 * pre-allocate all the necessary memory. In other words, it would never
		 * throw out of memory error.
		 */
		AllocateReservedConectionEntry(workerNode->workerName, workerNode->workerPort,
									   MyDatabaseId);

		/*
		 * Increment the shared counter, we may need to wait if there are
		 * no space left.
		 */
		WaitLoopForSharedConnection(workerNode->workerName, workerNode->workerPort);

		/* locally mark that we have one connection reserved */
		IncrementReservedConnection(workerNode->workerName, workerNode->workerPort,
									MyDatabaseId);
	}
}
```

We can move the `IsReservationRequired` to the caller as well.

Also I think `reserve shared connection` seems a bit better than `reserve shared connection counter`.

Reply:
The downside of the suggestion is that we'd iterate over the  workerNodeList twice, which might be OK for many cases.

I might adopt some parts of the suggestion, though probably avoiding multiple iterations considering that  we might expand  this code-path to all executions in the future

Reply:
You are right, there is that downside. But the list will be pretty small in most of the cases like maximum 10 elements maybe. There is a already a sort step so going over the list twice might be neglectable.

Reply:
checkout the latest changes. Would that be OK for you? Most importantly, the external API is fixed. The internal one is still `IfNeeded`, but that seems OK and it does what actually the name  suggests

Reply:
thanks for changing the API, I think the internal method is still doing a bit much as we haven't changed it. To apply the suggestion we should have an answer for this `How much performance regression will there be if we iterate the list twice instead of once?`. When we ignore the performance regression does the suggested version look better to you in terms of readability?

Reply:
> When we ignore the performance regression does the suggested version look better to you in terms of readability?

Yes, and I'm lazy for doing the performance test with say 3-10-20-50-100 nodes to have a good answer to the perf question :)

Reply:
I can do the performance test if necessary because I think there is a good improvement on the readability side.

Reply:
Most of the time loops are dominated by the contents of the loop which are repeated many times. Hence, two loops is usually not much slower than one loop. For instance, this loop currently does 3 hash lookups, which are probably 100x more expensive (each) than the pointer dereference that is required to iterate a linked list.